src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+            if opdata.block_height == ancestor_block && opdata.vtxindex == ancestor_vtxindex {
+                // found
+                debug!(
+                    "PoX anchor block-commit {},{},{} has {} burnt, {} confs",
+                    &opdata.txid, opdata.block_height, opdata.vtxindex, most_burnt, most_confs
+                );
+
+                let mut descendancy = Vec::with_capacity(prepare_ops.len());
+                for prepare_block_ops in prepare_ops.iter() {
+                    let mut block_descendancy = Vec::with_capacity(prepare_ops.len());
+                    for opdata in prepare_block_ops.iter() {
+                        if let Some((op_ancestor_height, op_ancestor_vtxindex, ..)) =
+                            ancestors.get(&(opdata.block_height, opdata.vtxindex))
+                        {
+                            if *op_ancestor_height == ancestor_block
+                                && *op_ancestor_vtxindex == ancestor_vtxindex
+                            {
+                                test_debug!("Block-commit {},{} descends from likely PoX anchor block {},{}", opdata.block_height, opdata.vtxindex, op_ancestor_height, op_ancestor_vtxindex);
+                                block_descendancy.push(true);
+                            } else {
+                                test_debug!("Block-commit {},{} does NOT descend from likely PoX anchor block {},{}", opdata.block_height, opdata.vtxindex, ancestor_block, ancestor_vtxindex);
+                                block_descendancy.push(false);
+                            }
+                        } else {
+                            test_debug!("Block-commit {},{} does NOT descend from likely PoX anchor block {},{}", opdata.block_height, opdata.vtxindex, ancestor_block, ancestor_vtxindex);
+                            block_descendancy.push(false);
+                        }
+                    }
+                    descendancy.push(block_descendancy);
+                }
+
+                return Ok(Some((opdata, descendancy)));
+            }
+        }
+    }
+
+    warn!("Evil miners confirmed a non-existant PoX anchor block!");
+    Ok(None)
+}
+
+/// Find the PoX anchor block selected in a reward cycle, if it exists.  This is the heaviest F*w-confirmed
+/// block-commit before the prepare-phase of this reward cycle, provided that it is not already an
+/// anchor block for some other reward cycle.  Note that the anchor block found will be the anchor
+/// block for the *next* reward cycle.
+/// Returns:
+///     (a) the list of block-commits, grouped by block and ordered by vtxindex, in this prepare phase
+///     (b) the PoX anchor block-commit, if it exists, and
+///     (c) the descendancy data for the prepare phase.  Descendency[i][j] is true if the jth
+///     block-commit in the ith block in the prepare phase descends from the anchor block, or False
+///     if not.
+pub fn find_pox_anchor_block<'a, B: BurnchainHeaderReader>(
```

Comment:
Does the PoX anchor block _also_ need to be the block confirmed by F*w _winning_ sortitions? This method only computes using the burnchain db (counting confirmations by any block commit). Or is this just returning a "candidate" anchor block?

Reply:
In Stacks 2.1, the PoX anchor block must not only be confirmed by F*w _winning_ sortitions, but also must correspond to the highest block-commit confirmed by the most BTC out of all block-commits with at least F*w descendants in the prepare phase.  In Stacks 2.0, only the first constraint is required.  If no such Stacks block is processed with both of these properties, then there is no anchor block in Stacks 2.1 rules.

* If miners are honest and always try to build off of the latest chain tip, then the only block-commit with this property will correspond to the anchor block.

* If miners broadcast block-commits that correspond to invalid blocks, then this code will still determine that there is a block-commit that corresponds to the anchor block, but its affirmation entry will be `a` since the chains coordinator won't select an anchor block.

* If through a stroke of bad luck two block-commits receive F*w confirming block-commits, but the block-commit confirmed by the most BTC does _not_ correspond to the Stacks block that was selected by the chains coordinator, then the affirmation entry will be `a` as well.